client-js/app/elements/labrad-registry.ts

```diff
@@ -202,226 +247,241 @@ export class LabradRegistry extends polymer.Base {
   }
 
-  @property({computed: '_computeSelectedType(selectedIdx)'})
-  selectedType: string;
 
-  @property({computed: '_computeSelectedItem(selectedIdx, kick)'})
-  selectedItem: string;
-
-  _computeSelectedType(selectedIdx: number): string {
+  private getSelectedType(): string {
     // Account for parent '..' entry.
     const offset = this.getListOffset();
-    if (this.dirs && selectedIdx < this.dirs.length + offset) {
+    if (this.dirs && this.getSelectedIndex() < this.dirs.length + offset) {
       return 'dir';
     } else {
       return 'key';
     }
   }
 
-  _computeSelectedItem(selectedIdx: number, kick: number): string {
-    // Account for parent '..' entry.
-    const offset = this.getListOffset();
-    const dirNames = (this.dirs || []).map(it => it.name);
-    const keyNames = (this.keys || []).map(it => it.name);
-    return dirNames.concat(keyNames)[selectedIdx - offset];
-  }
-
 
+  /**
+   * This is called whenever a newItem is added, hooked in actitives.
+   */
   async repopulateList(): Promise<void> {
-    var resp = await this.socket.dir({path: this.path});
+    const resp = await this.socket.dir({path: this.path});
+
     this.splice('dirs', 0, this.dirs.length);
     this.splice('keys', 0, this.keys.length);
-
-    for (var i in resp.dirs) {
-      this.push('dirs', {name: resp.dirs[i], url: this.places.registryUrl(resp.path, resp.dirs[i])});
-    }
-    for (var j in resp.keys) {
-      this.push('keys', {name: resp.keys[j], value: resp.vals[j]});
+    this.splice('listItems', 0, this.listItems.length);
+
+    if (this.path.length > 0 && this.places) {
+      const url = this.places.registryUrl(this.path.slice(0, -1));
+      this.push('listItems', {
+        name: '..',
+        isParent: true,
+        isDir: false,
+        isKey: false,
+        url: url
+      });
+    }
+
+    for (const name of resp.dirs) {
+      const url = this.places.registryUrl(resp.path, name);
+      this.push('dirs', {
+        name: name,
+        url: url,
+      });
+      this.push('listItems', {
+        name: name,
+        isParent: false,
+        isDir: true,
+        isKey: false,
+        url: url
+      });
+    }
+
+    for (const j in resp.keys) {
+      const name = resp.keys[j];
+      const value = resp.vals[j];
+      this.push('keys', {
+        name: name,
+        value: value,
+      });
+
+      this.push('listItems', {
+        name: name,
+        isParent: false,
+        isDir: false,
+        isKey: true,
+        value: value
+      });
+    }
+
+    const selected = this.selected;
+    this.applyFilterToListItems();
+
+    let index = this.getDefaultSelectedItem();
+
+    // If there was a previously selected item, say we were editing one, then
+    // we want to reselect it after the list is repopulated. To do this we
+    // fuzzy match the objects by requiring either the name or the value to be
+    // the same. This handles both the case of renaming a dir/key or changing
+    // the value of it.
```

Comment:
I don't see how this could work with renaming a dir, since there's no value to match, or at least none that uniquely identifies it compared to other dirs. I don't think we have a good way to fix that at the moment, but should change the comment to reflect the current behavior.